### ISSUE TITLE
Agent OOMException when preparing JUnit xml report

### ISSUE THREAD

**Issue by ketan** [gocd/gocd#2160]
##### Issue Type
- Bug Report
##### Summary

The agent runs out of memory when executing `UnitTestReportGenerator` with a fairly large number of XML test reports.
##### Steps to Reproduce

Pick a directory with a fairly large number of XML reports, I rsynced some over from build.go.cd. I had about 1600 files in there.

Call `UnitTestReportGenerator.generate()` with the directory containing these file and run with `-Xmx450m` (the agent runs with 512MB, so this feels "safe", but increasing the limit to 512MB does not help much because of the nature of the problem.  I tried with an Xmx of 2GB, and still hit this issue.
##### Expected Results

The agent should not OOM.
##### Log snippets

This is a stacktrace with go 15.2.0 and points to [this line](https://github.com/gocd/gocd/blob/15.2.0/config/config-api/src/com/thoughtworks/go/domain/UnitTestReportGenerator.java#L72). With recent versions, the line number will change, but the result is the same.

``` javastacktrace
at com.sun.org.apache.xalan.internal.xsltc.dom.XSLTCDTMManager.getDTM(XSLTCDTMManager.java:458)
at com.sun.org.apache.xalan.internal.xsltc.dom.XSLTCDTMManager.getDTM(XSLTCDTMManager.java:252)
at com.sun.org.apache.xalan.internal.xsltc.trax.TransformerImpl.getDOM(TransformerImpl.java:559)
at com.sun.org.apache.xalan.internal.xsltc.trax.TransformerImpl.transform(TransformerImpl.java:742)
at com.sun.org.apache.xalan.internal.xsltc.trax.TransformerImpl.transform(TransformerImpl.java:353)
at com.thoughtworks.go.domain.UnitTestReportGenerator.generate(UnitTestReportGenerator.java:72)
at com.thoughtworks.go.config.TestArtifactPlan.mergeAndUploadTestResult(TestArtifactPlan.java:97)
at com.thoughtworks.go.config.TestArtifactPlan.publish(TestArtifactPlan.java:62)
at com.thoughtworks.go.domain.DefaultJobPlan.publishArtifacts(DefaultJobPlan.java:105)
at com.thoughtworks.go.remote.work.BuildWork.completeJob(BuildWork.java:198)
at com.thoughtworks.go.remote.work.BuildWork.build(BuildWork.java:146)
at com.thoughtworks.go.remote.work.BuildWork.doWork(BuildWork.java:91)
at com.thoughtworks.go.agent.JobRunner.run(JobRunner.java:53)
at com.thoughtworks.go.agent.AgentController.retrieveWork(AgentController.java:186)
at com.thoughtworks.go.agent.AgentController.loop(AgentController.java:132)
```

It appears that the cause of this is the fact that the [XSL file](https://github.com/gocd/gocd/blob/15.2.0/common/resources/com/thoughtworks/go/domain/unittests.xsl) runs a whole lot of operations that [holds](https://github.com/gocd/gocd/blob/15.2.0/common/resources/com/thoughtworks/go/domain/unittests.xsl#L23) on to a lot of [nodes](https://github.com/gocd/gocd/blob/15.2.0/common/resources/com/thoughtworks/go/domain/unittests.xsl#L105-L107) as it processes the document.

Basically, unless we change _how_ this report is generated, no amounts of memory will solve the problem.

**Comment by ketan** [comment]
Fundamentally all that the XSLT does is to count the number of tests and collect all the tests with failures, something that we can do in a lot less computationally intensive manner using a simple SaxHandler.

I checked and processing most reports (except for the crazy nested [nunit reports](https://github.com/gocd/gocd/blob/cef918f39f513616be24fef34ec9c944ac193556/server/test/data/logs/log-dataset/nunit/nunit-tests-failed.xml)) should be fairly straight forward.

**Comment by ketan** [comment]
Ok then there's this gem — https://github.com/jenkinsci/xunit-plugin/tree/master/src/main/resources/org/jenkinsci/plugins/xunit/types :)

**Comment by arvindsv** [comment]
Related to this as well? https://github.com/gocd/gocd/issues/1132#issuecomment-166976539

**Comment by ketan** [comment]
> Related to this as well? #1132 (comment)

Not quite, but I suppose we could fix them both with a single "fix"

**Comment by ketan** [comment]
Agent 112 had this same exception today.

**Comment by ketan** [comment]
I'm not sure which of these exceptions could be the cause of the other.

<img width="1173" alt="screen shot 2016-06-05 at 12 44 58 pm" src="https://cloud.githubusercontent.com/assets/10598/15804108/5f3a7e6e-2b1b-11e6-85cf-65d46e4d5d00.png">
![Uploading Screen Shot 2016-06-05 at 12.45.13 PM.png…]()

**Comment by arvindsv** [comment]
Maybe `-XX:+HeapDumpOnOutOfMemoryError` will help?
